File: LeetCode/235. Lowest Common Ancestor of a Binary Search Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {//find the first node between p and q
        int min = Math.min(p.val, q.val);
        int max = Math.max(p.val, q.val);
        if (root.val >= min && root.val <= max) {
            return root;
        }
        if (root.val > max) {
            return lowestCommonAncestor(root.left, p, q);
        } else {
            return lowestCommonAncestor(root.right, p, q);
        }
    }
}

File: LeetCode/235. Lowest Common Ancestor of a Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        left, right = min(p.val, q.val), max(p.val, q.val)
        while root.val < left or root.val > right:
            if root.val < left:
                root = root.right
            else:
                root = root.left
        return root